### PR TITLE
Register candidate

### PR DESCRIPTION
**Pull request by JamBai2812**
Endpoint create for registering new candidates from a form submission (/candidates/register):

front end requires these keys for the form:
firstName
lastName
email

The API then generates a Guid for the user and writes the data to the candidates table in the database.
Guid is 36 character string, will need to be used by the 'generateURL' endpoint to create the unique link.

### CODE REVIEW

File: TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+{
+    public class CandidateRequest
+    {
+        public string FirstName { get; set; }
+        public string LastName { get; set; }
+        public string Email { get; set; }
+
+        public CandidateRequest()
+        {
+        }
```
Comment (MichaelWalker) [inline TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs]: You should just be able to delete this constructor, because its the same as the default one :)

File: TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace TestSwitchApi.Models.Response
+{
+    public class CandidateRegisteredResponse
+    {
+        public CandidateRegisteredResponse()
+        {
+        }
+    }
+}
```
Comment (MichaelWalker) [inline TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs]: Is this used anywhere?

File: TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs [hunk 3]
```diff
@@ -28,5 +30,18 @@ public int Count(PageRequest pageRequest)
...
+        {
+            var response = _context.Candidates.Add(new CandidateDataModel()
+            {
+                FirstName = candidateRequest.FirstName,
+                LastName = candidateRequest.LastName,
+                Email = candidateRequest.Email,
+                Guid = Guid.NewGuid().ToString(),
+            });
+            _context.SaveChanges();
+            return response.Entity;
```
Comment (MichaelWalker) [inline TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs:50]: so... some validation questions
- should we allow someone to sign up without a First Name / Last Name / Email?
- should we allow 2 people with the same Email?
- are their any restrictions on the length of any of these fields?

**Review by MichaelWalker — CHANGES_REQUESTED** [review]
We should probably think a little harder about validation (and how we test that)... but otherwise looks nice! :)

File: TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs (hunk 2 above)
Comment (JamBai2812) [reply TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs]: nope, forgot to delete

File: TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs (hunk 1 above)
Comment (JamBai2812) [reply TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs]: deleted

File: TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs (hunk 3 above)
Comment (JamBai2812) [reply TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs:50]: So yeah I thought about some similar validation that we did for the login page - ie. handle it with required={true}, is it better to validate on the server side too? And no probably should validate the email is unique, I'll have a look into that. Lengths are a problem in the DB so I'll carry those over to the validation in the end point!

Comment (MichaelWalker) [reply TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs:50]: cool - client side validation is just to allow us to give really quick, clear feedback to the user (ie its all about good UX).

it doesn't do anything to actually prevent the user doing something we don't want them to do - ie they could still do stuff in Postman, or by editing the JS or any number of other ways.

if we actually want to stop someone doing something, it MUST be done server side.

in this case we've got 2 lines of defense.
- we can validate the model: https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-3.1
- we can enforce constraints to the Database.

Validating the model gives nice clear error messages
Constraints on the DB are awesome at actually preventing stuff we don't want, but are likely to throw unclear error messages.

- for the email uniqueness - DB Unique constraint is nice and easy (can probably ignore nice error messages for now)
- making a field required I'd do both DB and Model Validation (`NOT NULL` for DB and `[Required]` on the model)
- length is already a restriction in the DB - adding matching validation to the model just helps us tell the user what they did wrong

File: TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs (hunk 1 above)
Comment (MichaelWalker) [reply TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs]: still seems to be here?

File: TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs (hunk 2 above)
Comment (MichaelWalker) [reply TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs]: still seems to be here?

File: TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs (hunk 1 above)
Comment (JamBai2812) [reply TestSwitchApi/TestSwitchApi/Models/Request/CandidateRequest.cs]: Oh I haven't pushed yet sorry

File: TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs (hunk 2 above)
Comment (JamBai2812) [reply TestSwitchApi/TestSwitchApi/Models/Response/CandidateRegisteredResponse.cs]: pushed now

File: TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs (hunk 3 above)
Comment (JamBai2812) [reply TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs:50]: Ok I think this is done now, only thing I would add is some unit tests, but I've done some manual testing in postman and seems to work! The DB already had all the correct restraints (NOT NULL for all columns and UNIQUE for email)

Comment (MichaelWalker) [reply TestSwitchApi/TestSwitchApi/Repositories/CandidatesRepo.cs:50]: nice! thanks

**Review by MichaelWalker — APPROVED** [review]
